### PR TITLE
feat., Imaginary: disable document processing by default

### PR DESCRIPTION
**Pull request by ernolf**
* Resolves: /nextcloud/server/issues/42175

## Summary

Document previews often appear as minimal, almost blank "miniatures" that can inadvertently reveal content, which may not always be desired.
Many users prefer having document icons instead of these previews.
The classic preview provider settings have considered this preference.
This feature reintroduces this choice for Imaginary.
Given that it is easier to generate previews than to remove existing ones, the default value is set to 'false'.

ernolf
## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: config/config.sample.php [hunk 1]
```diff
@@ -1307,6 +1307,17 @@
...
+/**
+ * Imaginary can, but does not, process PDF and Illustrator files by default.
+ * The following MIME types are always processed:
+ * bmp, x-bitmap, png, jpeg, gif, heic, heif, svg+xml, tiff and webp
+ * 
+ * Set to ``true``  to enable processing of documents of type pdf and illustrator.
+ * 
+ * Defaults to ``false``
+ */
+'imaginary_process_documents' => false,
```
Comment (szaimen) [inline config/config.sample.php:1319]:
```suggestion
 * Defaults to ``true``
 */
'imaginary_process_documents' => true,
```

File: lib/private/Preview/Imaginary.php [hunk 2]
```diff
@@ -40,7 +40,15 @@ public function getMimeType(): string {
 	}
 
 	public static function supportedMimeTypes(): string {
-		return '/(image\/(bmp|x-bitmap|png|jpeg|gif|heic|heif|svg\+xml|tiff|webp)|application\/(pdf|illustrator))/';
+		$config = \OC::$server->getConfig();
+		// Check if processing of documents is enabled
+		$processDocuments = $config->getSystemValueBool('imaginary_process_documents', false);
```
Comment (szaimen) [inline lib/private/Preview/Imaginary.php:45]:
```suggestion
		$processDocuments = $config->getSystemValueBool('imaginary_process_documents', true);
```

File: lib/private/Preview/Imaginary.php [hunk 3]
```diff
@@ -81,8 +89,17 @@ public function getCroppedThumbnail(File $file, int $maxX, int $maxY, bool $crop
...
 			case 'image/svg+xml':
+				$convert = true;
+				// Converted files do not need to be autorotated
+				$autorotate = false;
+				$mimeType = 'png';
+				break;
 			case 'application/pdf':
 			case 'application/illustrator':
+				// Check if processing of documents is enabled
+				if (!$this->config->getSystemValueBool('imaginary_process_documents', false)) {
```
Comment (szaimen) [inline lib/private/Preview/Imaginary.php:100]:
```suggestion
				if (!$this->config->getSystemValueBool('imaginary_process_documents', true)) {
```

**Review by szaimen — COMMENTED** [review]
Thanks for the PR @ernolf! 


I'd personally vote for enabling it by default and allowing to disable it

File: config/config.sample.php [hunk 4]
```diff
@@ -1307,6 +1307,17 @@
  */
 'preview_imaginary_key' => 'secret',
 
+/**
+ * Imaginary can, but does not, process PDF and Illustrator files by default.
```
Comment (szaimen) [inline config/config.sample.php:1311]:
```suggestion
 * Imaginary processes PDF and Illustrator files by default.
```

**Comment by kesselb** [comment]
Thanks for your pull request :+1: 

Cleaner, then a new configuration option, would be to use subclassing and reuse the existing preview provider configuration.

- Change Imaginary.supportedMimeTypes and remove the application mime type
- Create a new class ImaginaryDocuments extending Imaginary and overwrite supportedMimeTypes to return only the document mime types.
- To enable/disable the provider, add Imaginary or ImaginaryDocuments to the preview provider list. 



```diff
Index: lib/private/PreviewManager.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/private/PreviewManager.php b/lib/private/PreviewManager.php
--- a/lib/private/PreviewManager.php	(revision e6502b1399ae57a1ab3e9097ab5337e51216adb2)
+++ b/lib/private/PreviewManager.php	(date 1720727804062)
@@ -346,6 +346,7 @@
 		$this->registerCoreProvider(Preview\MP3::class, '/audio\/mpeg/');
 		$this->registerCoreProvider(Preview\OpenDocument::class, '/application\/vnd.oasis.opendocument.*/');
 		$this->registerCoreProvider(Preview\Imaginary::class, Preview\Imaginary::supportedMimeTypes());
+		$this->registerCoreProvider(Preview\ImaginaryDocuments::class, Preview\ImaginaryDocuments::supportedMimeTypes());
 
 		// SVG and Bitmap require imagick
 		if ($this->imagickSupport->hasExtension()) {
Index: lib/private/Preview/Imaginary.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/private/Preview/Imaginary.php b/lib/private/Preview/Imaginary.php
--- a/lib/private/Preview/Imaginary.php	(revision e6502b1399ae57a1ab3e9097ab5337e51216adb2)
+++ b/lib/private/Preview/Imaginary.php	(date 1720727664657)
@@ -40,7 +40,7 @@
 	}
 
 	public static function supportedMimeTypes(): string {
-		return '/(image\/(bmp|x-bitmap|png|jpeg|gif|heic|heif|svg\+xml|tiff|webp)|application\/(pdf|illustrator))/';
+		return '/(image\/(bmp|x-bitmap|png|jpeg|gif|heic|heif|svg\+xml|tiff|webp))/';
 	}
 
 	public function getCroppedThumbnail(File $file, int $maxX, int $maxY, bool $crop): ?IImage {
Index: lib/private/Preview/ImaginaryDocuments.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/private/Preview/ImaginaryDocuments.php b/lib/private/Preview/ImaginaryDocuments.php
new file mode 100644
--- /dev/null	(date 1720728758772)
+++ b/lib/private/Preview/ImaginaryDocuments.php	(date 1720728758772)
@@ -0,0 +1,15 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OC\Preview;
+
+class ImaginaryDocuments extends Imaginary {
+	public static function supportedMimeTypes(): string {
+		return '/application\/(pdf|illustrator)/';
+	}
+}
```

**Review by st3iny — COMMENTED** [review]
Conceptually, this seems like a good idea. 

However, I prefer kesselb's approach for its simplicity as this is mostly about changing supported mime types.

**Comment by ernolf** [comment]
> I prefer kesselb's approach for its simplicity as this is mostly about changing supported mime types.

I also believe @kesselb's suggestion is beneficial as it moves us towards standardizing preview handling.

Some thoughts:
When categorizing MIME types, we can group them as follows:

* **Images** (Lossy): jpeg, gif, webp
* **Raw Images** (Lossless): bmp, x-bitmap, png, heic, heif, tiff
* **Vector-based Images**: svg+xml, application/illustrator *(I have now categorized Illustrator among images, where it rightly belongs.)*
* **PDF/Documents**: application/pdf

Maintaining one provider per MIME type (as with legacy providers) or further dividing images into lossy, lossless, and vector-based categories might confuse existing Imaginary users. Therefore, I propose simplifying this into two categories: images processed by "Imaginary" and PDF document previews processed by "ImaginaryPDF". Since only PDF documents can be handled and Illustrator actually belongs among the images, introducing a provider called "ImaginaryDocuments" would also be rather confusing.

I've submitted a new pull request #46508 (this time as a member) that expands on kesselb's approach and this thoughts.

So, I kindly request your support there and would like to close this pull request if that would be OK for you.

**Comment by ernolf** [comment]
Since #46508 is merged, this can be closed

**Pull request closed without merging**